### ISSUE TITLE
Asynchronous scheduler for running tests concurrently

### ISSUE THREAD

**Issue by thilp** [darrenburns/ward#170]
Hello! TL;DR: I would like to talk about having alternatives to the existing "test runner" of Ward, and in particular a fully asynchronous alternative.

## My use-case

At work, we have an end-to-end test suite. It is written in Python (we use Pytest today) and it mostly interacts with external services (the Kubernetes API server, the REST/GraphQL APIs of the service under test, the GitHub API, you name it). Consequently, most of what the tests do is nothing, because they're waiting (polling) for some condition to become true in one of the external services.

That's why we run these tests in parallel, using `pytest-xdist`. But we're not very happy with `pytest-xdist`: the documentation is poor, the whole concept seems overkill for our use case (creating additional processes and workers, even though our tests are already completely async), and it behaves weirdly when you ask for not-so-many xdist workers in our EC2 instances.

Ideally, we would have a way to run _all_ our tests (not only the number of parallel workers xdist accepted to start) concurrently, in the same asyncio loop. Our tests are not unit tests, and they're not interested in running in a pristine, dedicated asyncio loop. What matters most to us is that they run as quickly as possible (since some scenarios can be long) while still checking what we want to check. Any issue caused by uncooperative tests would just be a bug that needs to be fixed; it would be on us devs, not on the test runner.

## The idea

I've read in #131:

>  this is as intended -- `async` tests are run sequentially. An `async` test runs asynchronous code on the event loop, but it waits until the test is complete before continuing on to the next test. So code **within** an async test is run asynchronously, but multiple `async` tests don't exist on the event loop at a given time (that would cause several issues).
> In other words, the tests themselves can be asynchronous, but the runner is not. It may be possible to have a separate asynchronous runner that users can opt-in to, but that'd be a looong way off likely :)

A quick (and probably imperfect) look at the code seems to indicate that `Suite.generate_test_runs` and `Test.run` is where the interesting scheduling part happens. The former sequentially calls the latter on all its Test instances. As far as I can see, if my use-case is something that makes sense for Ward, a pinch of dependency injection in these two functions with some kind of _SuiteRunner_ interface/protocol seems an obvious (if maybe naive) approach. Maybe some additional work for resolving fixtures too.

What do you think? Is the whole idea of an (opt-in) fully asynchronous test runner something you would be willing to see in Ward? Would you welcome further discussion and PRs in that direction? Do you have other suggestions that would make Ward suitable for this use-case? And if you're open to the idea, what do you think about the general technical solution I drafted above?

Thanks for reading and have a nice day! 😊

**Comment by darrenburns** [comment]
I'd love to see this, and I think the general approach you suggested seems reasonable. The approach you suggested is how I've been thinking of approaching a parallel test runner mode. I'd totally be open to PRs. I think it'd be a fantastic feature and would really help set Ward apart!

**Comment by thilp** [comment]
I'm really glad to hear that!

TL;DR: I have played a little bit with the code and come up with a limited prototype. The existing logic is so clear, it's a pleasure to work on it! I have two observations about challenges of layering sync and async code in Ward, and I'm still working on it.

### TestResult iterators: sync vs async

`Suite.generate_test_runs` returns an iterator of test results, but that is a _synchronous_ iterator; so my first attempt runs all tests concurrently, but can't return any test results (i.e. display anything to users) before _all_ tests are finished, because they run in a single `loop.run_until_complete` that can't exchange with the sync generator.

I couldn't find a way to "bridge" sync and async iterators, and I'm not sure that's possible with asyncio. The only solution to this I see for now is to make `generate_test_runs` (and probably some of its callers) async too. This "async contamination" upwards makes me a bit uncomfortable, because it sounds like a bigger refactoring than I initially thought. (I believe it would be a good idea for Ward regardless, but it looks like a larger feature.)

Getting all test results at once at the end is a UX nuisance, but it would still be acceptable for my use-case, so I carried on with my experiment.

### Async in sync in async (in sync)

Now, I'm at the stage where I have to resolve fixtures (either sync or async) for each test (either sync or async). I've kept as much as possible the existing functions, so for async fixtures I'm dealing with a sync/async layered cake like:
- `Suite.generated_test_runs` (still sync);
- some sync then async functions in my new "Scheduler" abstraction that I injected into Suite;
- `Test.run` (still sync) called from my async functions, using (depending on `Test.is_async_test`) a sync call or asyncio's `loop.run_in_executor`;
- TestArgumentResolver (still sync) calling `fixture.fn` (async).

Now for running async code from sync code already running in an async loop (still following? :upside_down_face:), I think I can only use `loop.create_task` and the like (as `loop.run_until_complete` expects a not-already-running loop). They return a task/future object, and I can't afford to await their completion because I'm in sync code at this point. (There's also `loop.call_soon` but I can't figure out how their _Handle_ return type is supposed to be useful.)

So it looks like I'll have to async-contaminate (or duplicate) a lot of `Test.run` and TestArgumentResolver logic if I want to be able to have async fixtures in concurrent async tests. I'm still looking for alternatives and will continue experimenting around that.

If you've read so far, I'm happy to hear any feedback! I'd also like to clarify that this is an (ongoing) experiment to see how far I can push the current design of Ward to get what I want; I don't have any expectations that anything I did (or talked about here) will make it into stable Ward, especially if "asyncio everywhere" is not on its roadmap. At least I hope my attempts can help if Ward decides to use more async in the future. (At best I'm still eyeing a concurrent test runner :grin:)

I expect to share more findings (or a conclusion) in the coming days.

**Comment by darrenburns** [comment]
Hey @thilp! Just wondering if you experimented on this any further at all? I've been rather negligent of Ward throughout the pandemic but am currently ramping up a little bit on it again recently, and am looking to take stock of where we're at 😄.
